Support extracting PubMed IDs from PubMed URL

konieczkow opened the pull request:
Why:

We were only extracting PubMed IDs if they were surrounded by spaces.
This was to reduce the number of false positives (extracting random numbers from DOIs for example).
The side effect of that change was that we would miss totally valid PubMed IDs coming from PubMed URLs.

This change addresses the need by:

This adds a separate extraction logic for extracting PubMed IDs from PubMed URLs.

Fixes #19

konieczkow committed:
Support extracting PubMed IDs from PubMed URL

Why:

We were only extracting PubMed IDs if they were surrounded by spaces.
This was to reduce the number of false positives (extacting random
numbers from DOIs for example).
The side effect of that change was that we would miss totally valid
PubMed IDs coming from PubMed URLs.

This change addresses the need by:

This adds a separate extraction logic for extracting PubMed IDs from
PubMed URLs.

jbilbo commented on lib/identifiers/pubmed_id.rb:
```diff
@@ -1,7 +1,40 @@
+    end
+
+    def pubmed_id_from_url
+      pubmed_candidate.sub(URL_REGEX, '')
```
Shall we also take into account any possible query/hash parameters? e.g. `https://www.ncbi.nlm.nih.gov/pubmed/6191871#cm6191871_69589`

jbilbo committed:
Skip query params and support multiple occurrences

Use `scan` to retrieve all PMIDs occurrences from URLs. Also, avoid
all query/hash parameters the URL might have.

mudge commented on lib/identifiers/pubmed_id.rb:
```diff
@@ -1,7 +1,9 @@
     def self.extract(str)
-      str.to_s.scan(/(?<=^|[[:space:]])0*(?!0)(\d+)(?=$|[[:space:]])/).flatten
+      str = str.to_s
+      str.scan(/(?<=^|[[:space:]])0*(?!0)(\d+)(?=$|[[:space:]])/).flatten |
```
Perhaps we could extract these patterns into named constants to explain the two types we support?

```ruby
ZERO_PADDED_NUMBER = ...
PUBMED_URL = ...
```

mudge commented on lib/identifiers/pubmed_id.rb:
```diff
@@ -1,7 +1,9 @@
-      str.to_s.scan(/(?<=^|[[:space:]])0*(?!0)(\d+)(?=$|[[:space:]])/).flatten
+      str = str.to_s
+      str.scan(/(?<=^|[[:space:]])0*(?!0)(\d+)(?=$|[[:space:]])/).flatten |
+        str.scan(%r{https?://(?:www\.)?ncbi.nlm.nih.gov/pubmed/(\d+)}).flatten
```
Does this mean we wouldn't strip zero-padding from the front of a PubMed ID in a URL? Perhaps we could lift the domain prefix here into the first non-capturing group of the previous pattern? Due to restrictions in look-behinds, we might need to change it like so:

```ruby
str.scan(%r{(?:^|[[:space:]]|https?://(?:www\.)?ncbi\.nlm\.nih\.gov/pubmed/)0*(?!0)(\d+)(?=$|[[:space:]])}).flatten
```

mudge replied:
This might then benefit from being a long-form pattern instead, e.g.

```ruby
%r{
  (?: # Valid prefixes for PubMed IDs
    ^ | # Start of the string
    [[:space:]] | # Leading whitespace
    https?://(?:www\.)?ncbi\.nlm\.nih\.gov/pubmed/ # A PubMed URL
  )
  0* # Any number of leading zeroes
  (?!0)(\d+) # A number not beginning with a zero
  (?=$|[[:space:]]) # Either the end of the string or trailing spaces
}x
```

jbilbo replied:
Since they're different approaches with different needs I don't think they match well when we join them together. The Pubmed URL don't really cares about how the string ends: After the Pubmed ID we could have `spaces`, `/`, `?`, `#`, `$`... 
But on the other way, it's important for the zero-padded-integer extraction. We don't want to match integers ending with `?` or `#` for example.

I've added zero-leading support to the Pubmed URL but they remain separate.

jbilbo committed:
Support leading zeroes in Pubmed URLs

We can extract Pubmed IDs with leading zeroes
inside Pubmed URLs.
The Regexps have been extracted into named constants to explain the two
types we support.

mudge commented on lib/identifiers/pubmed_id.rb:
```diff
@@ -1,7 +1,11 @@
 module Identifiers
   class PubmedId
+    ZERO_PADDED_NUMBER = %r{(?<=^|[[:space:]])0*(?!0)(\d+)(?=$|[[:space:]])}
+    PUBMED_URL = %r{https?://(?:www\.)?ncbi\.nlm\.nih\.gov/pubmed/0*(\d+)}i
```
Should we also support mobile PubMed URLs, e.g. https://www.ncbi.nlm.nih.gov/m/pubmed/22497409?

jbilbo committed:
Add support to mobile PubMed URLs
